ykcompile/src/lib.rs

```diff
@@ -222,6 +234,63 @@ impl TraceCompiler {
         Ok(())
     }
 
+    /// Compile a call to a native symbol. This is used for occasions where you don't want the
+    /// callee inlined, or cannot inline it (e.g. it's a foreign function).
+    fn c_call(
+        &mut self,
+        opnd: &CallOperand,
+        args: &Vec<Operand>,
+        dest: &Option<Place>,
+    ) -> Result<(), CompileError> {
+        let sym = if let CallOperand::Fn(sym) = opnd {
+            sym
+        } else {
+            return Err(CompileError::Unimplemented(
+                "unknown call target".to_owned(),
+            ));
+        };
+
+        // FIXME implement support for arguments.
+        if !args.is_empty() {
+            return Err(CompileError::Unimplemented("call with args".to_owned()));
+        }
+
+        // Generally speaking, we don't know what the callee will clobber, so we save our registers
+        // before doing the call. For now we assume that the caller is OK with RAX being clobbered
+        // though.
+        for reg in 1..X86_86_NUM_REGS {
+            dbg!(reg);
+            dynasm!(self.asm
+                ; push Rq(reg)
+            );
+        }
+
+        let sym_addr = TraceCompiler::find_symbol(sym)? as i64;
+        dynasm!(self.asm
+            ; xor rax, rax
```

Comment:
Since the `call` will overwrite `rax` do we really need to clear it beforehand?

Reply:
In the Sys-V calling convention, `rax` (well, part of it) specifies the number of vector register args for a varag call.

Snippet from sysv ABI spec:
![image](https://user-images.githubusercontent.com/604955/82880998-a4c66980-9f2e-11ea-8579-bf7bfba433a0.png)

https://uclibc.org/docs/psABI-x86_64.pdf

We can't handle arguments (let alone varags) yet, but I thought it'd be wise to zero the "hidden argument" to avoid surprises later.

I remember a time during my PhD when I spent half a day figuring out why I couldn't call `printf` from asm. It was because `rax` contained left-over junk.

Perhaps I should add a comment?

Reply:
Ah, that makes sense then. I guess a short comment can't hurt though.